File: ios/Talise/Features/Chat/ChatHistorySheet.swift
import SwiftUI

/// A little history — recent chats in a compact medium-detent sheet (not a
/// full-screen sidebar). Tap a row to reopen it; "New" starts fresh; swipe-free
/// delete on each row.
struct ChatHistorySheet: View {
    let vm: ChatViewModel
    var onPick: () -> Void

    @State private var query = ""

    private var filtered: [ChatConversation] {
        let q = query.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !q.isEmpty else { return vm.conversations }
        return vm.conversations.filter { $0.title.localizedCaseInsensitiveContains(q) }
    }

    var body: some View {
        VStack(spacing: 0) {
            HStack {
                Text("Chats")
                    .font(TaliseFont.heading(20, weight: .semibold))
                    .foregroundStyle(TaliseColor.fg)
                Spacer()
                Button { vm.newChat(); onPick() } label: {
                    HStack(spacing: 6) {
                        Image(systemName: "square.and.pencil").font(.system(size: 13, weight: .medium))
                        Text("New").font(TaliseFont.body(14, weight: .medium))
                    }
                    .foregroundStyle(TaliseColor.bg)
                    .padding(.horizontal, 14).padding(.vertical, 8)
                    .background(Capsule().fill(TaliseColor.greenMint))
                }
                .buttonStyle(.plain)
            }
            .padding(.horizontal, 20)
            .padding(.top, 20)
            .padding(.bottom, 12)

            // Search — filter past chats by title.
            if !vm.conversations.isEmpty {
                HStack(spacing: 8) {
                    Image(systemName: "magnifyingglass")
                        .font(.system(size: 13))
                        .foregroundStyle(TaliseColor.fgDim)
                    TextField("Search chats", text: $query)
                        .font(TaliseFont.body(14))
                        .foregroundStyle(TaliseColor.fg)
                        .textInputAutocapitalization(.never)
                        .autocorrectionDisabled()
                    if !query.isEmpty {
                        Button { query = "" } label: {
                            Image(systemName: "xmark.circle.fill")
                                .font(.system(size: 14))
                                .foregroundStyle(TaliseColor.fgDim)
                        }
                        .buttonStyle(.plain)
                    }
                }
                .padding(.horizontal, 14).padding(.vertical, 10)
                .background(RoundedRectangle(cornerRadius: 12, style: .continuous).fill(TaliseColor.surface))
                .padding(.horizontal, 20).padding(.bottom, 10)
            }

            if vm.conversations.isEmpty {
                Spacer()
                VStack(spacing: 8) {
                    Image(systemName: "clock")
                        .font(.system(size: 28, weight: .light))
                        .foregroundStyle(TaliseColor.fgDim)
                    Text("No past chats yet.")
                        .font(TaliseFont.body(14, weight: .light))
                        .foregroundStyle(TaliseColor.fgMuted)
                }
                Spacer()
                Spacer()
            } else {
                ScrollView(showsIndicators: false) {
                    VStack(spacing: 2) {
                        ForEach(filtered) { c in row(c) }
                        if filtered.isEmpty {
                            Text("No chats match that search.")
                                .font(TaliseFont.body(13, weight: .light))
                                .foregroundStyle(TaliseColor.fgMuted)
                                .padding(.top, 24)
                        }
                    }
                    .padding(.horizontal, 12)
                    .padding(.bottom, 8)
                    // Trust note: deleting a chat clears the local transcript only.
                    // The durable memory lives on Walrus (append-only) and stays.
                    Text("Deleting a chat only clears it from here. What Talise has learned stays saved on Walrus.")
                        .font(TaliseFont.mono(10, weight: .light))
                        .foregroundStyle(TaliseColor.fgDim)
                        .multilineTextAlignment(.center)
                        .padding(.horizontal, 28).padding(.top, 10).padding(.bottom, 16)
                }
            }
        }
        .background(TaliseColor.bg)
    }

    private func row(_ c: ChatConversation) -> some View {
        HStack(spacing: 10) {
            Button { vm.open(c.id); onPick() } label: {
                VStack(alignment: .leading, spacing: 2) {
                    Text(c.title.isEmpty ? "New chat" : c.title)
                        .font(TaliseFont.body(15, weight: .regular))
                        .foregroundStyle(TaliseColor.fg)
                        .lineLimit(1)
                    Text(relativeTime(c.updatedAt))
                        .font(TaliseFont.mono(10, weight: .light))
                        .foregroundStyle(TaliseColor.fgDim)
                }
                .frame(maxWidth: .infinity, alignment: .leading)
                .contentShape(Rectangle())
            }
            .buttonStyle(.plain)

            Button { vm.deleteConversation(c.id) } label: {
                Image(systemName: "trash")
                    .font(.system(size: 12))
                    .foregroundStyle(TaliseColor.fgDim)
                    .frame(width: 30, height: 30)
            }
            .buttonStyle(.plain)
            .accessibilityLabel("Delete chat")
        }
        .padding(.horizontal, 14)
        .padding(.vertical, 11)
        .background(RoundedRectangle(cornerRadius: 14, style: .continuous).fill(TaliseColor.surface))
    }

    private func relativeTime(_ d: Date) -> String {
        let f = RelativeDateTimeFormatter()
        f.unitsStyle = .abbreviated
        return f.localizedString(for: d, relativeTo: Date())
    }
}

File: android/app/src/main/java/io/talise/app/feature/chat/ChatHistorySheet.kt
package io.talise.app.feature.chat

import android.text.format.DateUtils
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Cancel
import androidx.compose.material.icons.filled.Edit
import androidx.compose.material.icons.filled.Search
import androidx.compose.material.icons.outlined.Delete
import androidx.compose.material.icons.outlined.Schedule
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import io.talise.app.ui.theme.TaliseColors
import io.talise.app.ui.theme.TaliseType

/**
 * A little history — recent chats in a compact sheet (not a full-screen
 * sidebar). Tap a row to reopen it; "New" starts fresh; per-row delete. The
 * Android port of iOS `ChatHistorySheet`.
 */
@Composable
fun ChatHistorySheet(
    conversations: List<ChatConversation>,
    onNew: () -> Unit,
    onOpen: (String) -> Unit,
    onDelete: (String) -> Unit,
) {
    var query by remember { mutableStateOf("") }
    val filtered = remember(conversations, query) {
        val q = query.trim()
        if (q.isEmpty()) conversations
        else conversations.filter { it.title.contains(q, ignoreCase = true) }
    }

    Column(Modifier.fillMaxWidth().background(TaliseColors.bg)) {
        Row(
            Modifier
                .fillMaxWidth()
                .padding(horizontal = 20.dp)
                .padding(top = 20.dp, bottom = 12.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Text("Chats", style = TaliseType.heading(20.sp, FontWeight.SemiBold), color = TaliseColors.fg)
            Spacer(Modifier.weight(1f))
            Row(
                Modifier
                    .clip(RoundedCornerShape(18.dp))
                    .background(TaliseColors.greenMint)
                    .clickable { onNew() }
                    .padding(horizontal = 14.dp, vertical = 8.dp),
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(6.dp),
            ) {
                Icon(Icons.Filled.Edit, contentDescription = null, tint = TaliseColors.bg, modifier = Modifier.size(14.dp))
                Text("New", style = TaliseType.body(14.sp, FontWeight.Medium), color = TaliseColors.bg)
            }
        }

        // Search — filter past chats by title.
        if (conversations.isNotEmpty()) {
            Row(
                Modifier
                    .fillMaxWidth()
                    .padding(horizontal = 20.dp)
                    .padding(bottom = 10.dp)
                    .clip(RoundedCornerShape(12.dp))
                    .background(TaliseColors.surface)
                    .padding(horizontal = 14.dp, vertical = 10.dp),
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(8.dp),
            ) {
                Icon(Icons.Filled.Search, contentDescription = null, tint = TaliseColors.fgDim, modifier = Modifier.size(15.dp))
                Box(Modifier.weight(1f)) {
                    if (query.isEmpty()) {
                        Text("Search chats", style = TaliseType.body(14.sp), color = TaliseColors.fgDim)
                    }
                    BasicTextField(
                        value = query,
                        onValueChange = { query = it },
                        singleLine = true,
                        textStyle = TaliseType.body(14.sp).copy(color = TaliseColors.fg),
                        cursorBrush = SolidColor(TaliseColors.accent),
                        modifier = Modifier.fillMaxWidth(),
                    )
                }
                if (query.isNotEmpty()) {
                    Icon(
                        Icons.Filled.Cancel,
                        contentDescription = "Clear search",
                        tint = TaliseColors.fgDim,
                        modifier = Modifier.size(16.dp).clickable { query = "" },
                    )
                }
            }
        }

        if (conversations.isEmpty()) {
            Column(
                Modifier.fillMaxWidth().padding(vertical = 64.dp),
                horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.spacedBy(8.dp),
            ) {
                Icon(Icons.Outlined.Schedule, contentDescription = null, tint = TaliseColors.fgDim, modifier = Modifier.size(28.dp))
                Text("No past chats yet.", style = TaliseType.body(14.sp, FontWeight.Light), color = TaliseColors.fgMuted)
            }
        } else {
            LazyColumn(
                Modifier.fillMaxWidth().weight(1f, fill = false),
                contentPadding = androidx.compose.foundation.layout.PaddingValues(horizontal = 12.dp),
                verticalArrangement = Arrangement.spacedBy(2.dp),
            ) {
                items(filtered, key = { it.id }) { c ->
                    HistoryRow(c, onOpen = { onOpen(c.id) }, onDelete = { onDelete(c.id) })
                }
                if (filtered.isEmpty()) {
                    item {
                        Text(
                            "No chats match that search.",
                            style = TaliseType.body(13.sp, FontWeight.Light),
                            color = TaliseColors.fgMuted,
                            textAlign = TextAlign.Center,
                            modifier = Modifier.fillMaxWidth().padding(top = 24.dp),
                        )
                    }
                }
                item {
                    // Trust note: deleting a chat clears the local transcript only.
                    // The durable memory lives on Walrus (append-only) and stays.
                    Text(
                        "Deleting a chat only clears it from here. What Talise has learned stays saved on Walrus.",
                        style = TaliseType.mono(10.sp, FontWeight.Light),
                        color = TaliseColors.fgDim,
                        textAlign = TextAlign.Center,
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(horizontal = 28.dp)
                            .padding(top = 10.dp, bottom = 16.dp),
                    )
                }
            }
        }
    }
}

@Composable
private fun HistoryRow(c: ChatConversation, onOpen: () -> Unit, onDelete: () -> Unit) {
    Row(
        Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(14.dp))
            .background(TaliseColors.surface)
            .padding(horizontal = 14.dp, vertical = 11.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        Column(
            Modifier.weight(1f).clickable { onOpen() },
            verticalArrangement = Arrangement.spacedBy(2.dp),
        ) {
            Text(
                c.title.ifEmpty { "New chat" },
                style = TaliseType.body(15.sp),
                color = TaliseColors.fg,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
            )
            Text(relativeTime(c.updatedAtMs), style = TaliseType.mono(10.sp, FontWeight.Light), color = TaliseColors.fgDim)
        }
        Box(
            Modifier.size(30.dp).clickable { onDelete() },
            contentAlignment = Alignment.Center,
        ) {
            Icon(Icons.Outlined.Delete, contentDescription = "Delete chat", tint = TaliseColors.fgDim, modifier = Modifier.size(15.dp))
        }
    }
}

private fun relativeTime(ms: Long): String =
    DateUtils.getRelativeTimeSpanString(ms, System.currentTimeMillis(), DateUtils.MINUTE_IN_MILLIS).toString()
